### PR TITLE
Issue according to signing preferences

### PR DESCRIPTION
**Pull request by lamara**
Dependent on this PR https://github.com/digitalbazaar/bedrock-identity/pull/11

### CODE REVIEW

File: lib/issuer.js [hunk 1]
```diff
@@ -7,11 +7,15 @@
  */
 var async = require('async');
 var bedrock = require('bedrock');
+var brIdentity = require('bedrock-identity');
+var brKey = require('bedrock-key');
 var brPermission = require('bedrock-permission');
+var cors = require('cors');
```
Comment (davidlehn) [inline lib/issuer.js]: cors and rest are not used.

File: package.json [hunk 2]
```diff
@@ -25,6 +25,7 @@
   "homepage": "https://github.com/digitalbazaar/bedrock-issuer",
   "dependencies": {
     "async": "^1.5.2",
+    "cors": "^2.7.1",
```
Comment (davidlehn) [inline package.json]: is cors needed?

File: lib/config.js [hunk 3]
```diff
@@ -77,10 +43,17 @@ roles['bedrock.credential.issuer'] = {
   comment: 'Role for credential issuers.',
   sysPermission: [
     permissions.CREDENTIAL_INSERT.id,
-    permissions.CREDENTIAL_ISSUE.id
+    permissions.CREDENTIAL_ISSUE.id,
+    permissions.PUBLIC_KEY_ACCESS.id,
```
Comment (dlongley) [inline lib/config.js]: Remove ending comma.

File: lib/issuer.js [hunk 4]
```diff
@@ -123,15 +116,18 @@ api.issue = function(actor, credential, options, callback) {
...
+    sign: ['getSigningKey', function(callback, results) {
+      if(!results.getSigningKey) {
         return callback(null, credential);
       }
       // digitally-sign credential
       jsigs.sign(credential, {
         algorithm: 'LinkedDataSignature2015',
-        privateKeyPem: options.privateKey,
-        creator: options.publicKeyId
+        privateKeyPem: results.getSigningKey.privateKey.privateKeyPem,
```
Comment (dlongley) [inline lib/issuer.js:130]: Add TODO for passing a signing function instead when supported by jsigs.

File: lib/issuer.js [hunk 5]
```diff
@@ -184,10 +176,13 @@ api.revoke = function(actor, id, options, callback) {
...
         credential.revocationReason = options.reason;
       }
+      if(!results.getSigningKey) {
+        return callback(null, credential);
+      }
       // digitally-sign credential
       jsigs.sign(credential, {
-        privateKeyPem: options.privateKey,
-        creator: options.publicKeyId
+        privateKeyPem: results.getSigningKey.privateKey.privateKeyPem,
```
Comment (dlongley) [inline lib/issuer.js:188]: Add TODO for passing a signing function instead when supported by jsigs.

File: lib/issuer.js [hunk 6]
```diff
@@ -245,6 +240,59 @@ api.claim = function(actor, id, callback) {
...
+      });
+    }],
+    getSigningKey: ['getPreferredSigningKey', function(callback, results) {
+      if(results.getPreferredSigningKey) {
+        // Use the preferred signing key
+        return callback(null, results.getPreferredSigningKey);
+      }
+      var identity = results.getIssuer;
+      // If no preferred signing key, get their first non-revoked public key
+      brKey.getPublicKeys(identity.id, identity, function(err, results) {
```
Comment (dlongley) [inline lib/issuer.js]: Should we do this or throw an error that says no preferred signing key?

Comment (dlongley) [inline lib/issuer.js]: If no preferred key, raise an error.

Comment (dlongley) [inline lib/issuer.js]: Need to add an "action" for setting the preferred key to the keys view for key rows.

**Comment by lamara** [comment]
Addressed comments, waiting on dependent PRs to get merged in now.

File: lib/issuer.js [hunk 7]
```diff
@@ -245,6 +244,42 @@ api.claim = function(actor, id, callback) {
...
+      if(!keyId) {
+        return callback();
+      }
+      brKey.getPublicKey({id: keyId}, identity,
+        function(err, publicKey, meta, privateKey) {
+        publicKey.privateKey = privateKey;
+        callback(err, publicKey);
+      });
+    }]
+  }, function(err, results) {
```
Comment (davidlehn) [inline lib/issuer.js:271]: Not sure about the error handling here. All errors above will be discarded and below error used instead. May want to at least pass along `err` as the final cause argument.  That would help in debugging if `brIdentity.get` or `brKey.getPublicKey` resulted in an error.

File: lib/issuer.js [hunk 8]
```diff
@@ -245,6 +244,42 @@ api.claim = function(actor, id, callback) {
...
+    if(!results.getPreferredSigningKey) {
+      return callback(new BedrockError(
+        'Preferred credential signing key not found.',
+        'NotFound',
+        {httpStatusCode: 404, public: true})
+      );
```
Comment (dlongley) [inline lib/issuer.js:277]: Add 4th param (I believe) which is the `cause` by passing `err`.

File: lib/issuer.js (hunk 4 above)
Comment (dlongley) [inline lib/issuer.js:130]: May have missed this one.

Comment (lamara) [inline lib/issuer.js:130]: It's there! the last commit isn't getting shown in the overview for whatever reason, but if you go to Files Changed you'll see it.

File: package.json [hunk 9]
```diff
@@ -35,7 +35,7 @@
     "bedrock-credentials-mongodb": "^2.0.0",
     "bedrock-credentials-rest": "^0.3.0",
     "bedrock-express": "^2.0.0",
-    "bedrock-identity": "^4.0.0",
+    "bedrock-identity": "digitalbazaar/bedrock-identity#modelFields",
```
Comment (dlongley) [inline package.json]: Needs to be updated to new release of bedrock-identity.

File: tests/test.js [hunk 10]
```diff
@@ -196,6 +226,27 @@ describe('bedrock-issuer', function() {
       req.end(postData);
     });
 
+    it('should return 400 if issue has no keys', function(done) {
```
Comment (dlongley) [inline tests/test.js]: Typo 'issue' => 'issuer'

File: package.json (hunk 9 above)
Comment (dlongley) [inline package.json]: Should be `"^4.2.0"`

**Comment by dlongley** [comment]
Two minor comments to address, then reassign to @mattcollier for merging.

**Comment by dlongley** [comment]
@mattcollier -- go ahead and release v1.0.0 after merge.

**Pull request merged**